scylla/tests/integration/caching_session.rs

```diff
@@ -0,0 +1,140 @@
+                .map(|_| mpsc::unbounded_channel::<(RequestFrame, Option<TargetShard>)>())
+                .unzip();
+            for (i, tx) in feedback_txs.iter().cloned().enumerate() {
+                running_proxy.running_nodes[i].change_request_rules(Some(vec![RequestRule(
+                    Condition::and(Condition::RequestOpcode(RequestOpcode::Prepare),Condition::not(Condition::ConnectionRegisteredAnyEvent)),
+                    RequestReaction::noop().with_feedback_when_performed(tx),
+                )]));
+            }
+
+
+            session.ddl(format!("CREATE KEYSPACE IF NOT EXISTS {} WITH REPLICATION = {{'class' : 'NetworkTopologyStrategy', 'replication_factor' : 1}}", ks)).await.unwrap();
+            session.use_keyspace(ks, false).await.unwrap();
+            session
+                .ddl("CREATE TABLE IF NOT EXISTS tab (a int, b int, c int, primary key (a, b, c))")
+                .await
+                .unwrap();
+            // Assumption: all nodes have the same number of shards
+            let nr_shards = match session.get_cluster_state().get_nodes_info().first() {
+                Some(node_info) => match node_info.sharder() {
+                    Some(sharder) => sharder.nr_shards,
+                    None => NonZero::new(1).unwrap(), // If there is no sharder, assume 1 shard
+                },
+                None => panic!("No nodes information available"),
+            };
+
+            // Consume all feedbacks so far to ensure we will not count something unrelated.
+            let _feedbacks: usize = feedback_rxs.iter_mut().map(consume_current_feedbacks).sum();
+
+            let caching_session: CachingSession = CachingSession::from(session, 100);
+            let insert_1_b_c = "INSERT INTO tab (a, b, c) VALUES (1, ?, ?)";
+            let insert_2_b_c = "INSERT INTO tab (a, b, c) VALUES (2, ?, ?)";
+            let insert_3_b_c = "INSERT INTO tab (a, b, c) VALUES (3, ?, ?)";
+            let insert_4_b_c = "INSERT INTO tab (a, b, c) VALUES (4, ?, ?)";
+            let mut batch = Batch::new(BatchType::Logged);
+            batch.append_statement(insert_1_b_c);
+            batch.append_statement(insert_2_b_c);
+            batch.append_statement(insert_3_b_c);
+            batch.append_statement(insert_4_b_c);
+            // First batch that should generate prepares for each shard.
+            let batch_values = vec![(1, 1), (2, 2), (3, 3), (4, 4)];
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+            // Few extra runs. Those batches should not result in any prepares being sent.
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
```

Comment:
:pick: Let's use `for` loop instead of repetition here.

Reply:
Switched to for loops from 1 to `=batch_size`